From 8d7d531c11e0c5986446ccdcb70a63044b60a46a Mon Sep 17 00:00:00 2001
From: ouyanghui <ouyanghui@eswincomputing.com>
Date: Wed, 17 Sep 2025 14:19:02 +0800
Subject: [PATCH 443/540] WIN2030-19444:fix:cpu core voltage boost causes
 system crashes.

Changelogs:
1.clock-contriller add cpu opp tables
2.deepcomputing core voltage keep 0.9v

Change-Id: I1309e651e861cb1b15924d0934856b72252d265b
Signed-off-by: ouyanghui <ouyanghui@eswincomputing.com>
---
 .../eic7702-deepcomputing-fml13v03-evt.dts    | 28 ++++++++++++++---
 .../eswin/eic7702-deepcomputing-fml13v03.dts  | 30 +++++++++++++++----
 drivers/clk/eswin/clk-eic7700.c               |  9 ++++++
 drivers/clk/eswin/clk.c                       | 27 ++++++++---------
 drivers/clk/eswin/clk.h                       |  5 ++--
 5 files changed, 72 insertions(+), 27 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts
index 405386d0866d..6cc7838a2619 100755
--- a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts
+++ b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03-evt.dts
@@ -261,13 +261,33 @@ d1_zero_device: zero-device@3a000000 {
 };
 
 &d0_cpu_opp_table {
-	/delete-node/ opp-1200000000;
-	/delete-node/ opp-1300000000;
+	opp-700000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-800000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-900000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-1400000000 {
+		opp-microvolt = <900000>;
+	};
 };
 
 &d1_cpu_opp_table {
-	/delete-node/ opp-1200000000;
-	/delete-node/ opp-1300000000;
+	opp-700000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-800000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-900000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-1400000000 {
+		opp-microvolt = <900000>;
+	};
 };
 
 &d0_clock {
diff --git a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts
index dbba03791ad3..75001119d7cd 100644
--- a/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts
+++ b/arch/riscv/boot/dts/eswin/eic7702-deepcomputing-fml13v03.dts
@@ -277,13 +277,22 @@ &d0_cpu_opp_table {
 	/delete-node/ opp-400000000;
 	/delete-node/ opp-500000000;
 	/delete-node/ opp-600000000;
-	/delete-node/ opp-700000000;
-	/delete-node/ opp-800000000;
-	/delete-node/ opp-900000000;
+	opp-700000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-800000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-900000000 {
+		opp-microvolt = <900000>;
+	};
 	/delete-node/ opp-1000000000;
 	/delete-node/ opp-1100000000;
 	/delete-node/ opp-1200000000;
 	/delete-node/ opp-1300000000;
+	opp-1400000000 {
+		opp-microvolt = <900000>;
+	};
 };
 
 &d1_cpu_opp_table {
@@ -294,13 +303,22 @@ &d1_cpu_opp_table {
 	/delete-node/ opp-400000000;
 	/delete-node/ opp-500000000;
 	/delete-node/ opp-600000000;
-	/delete-node/ opp-700000000;
-	/delete-node/ opp-800000000;
-	/delete-node/ opp-900000000;
+	opp-700000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-800000000 {
+		opp-microvolt = <900000>;
+	};
+	opp-900000000 {
+		opp-microvolt = <900000>;
+	};
 	/delete-node/ opp-1000000000;
 	/delete-node/ opp-1100000000;
 	/delete-node/ opp-1200000000;
 	/delete-node/ opp-1300000000;
+	opp-1400000000 {
+		opp-microvolt = <900000>;
+	};
 };
 
 &d0_clock {
diff --git a/drivers/clk/eswin/clk-eic7700.c b/drivers/clk/eswin/clk-eic7700.c
index ee01dd2d1890..dc4e945f7fce 100644
--- a/drivers/clk/eswin/clk-eic7700.c
+++ b/drivers/clk/eswin/clk-eic7700.c
@@ -12,6 +12,8 @@
 #include <linux/of.h>
 #include <linux/of_device.h>
 #include <linux/clk.h>
+#include <linux/devfreq.h>
+#include <linux/pm_opp.h>
 #include <dt-bindings/clock/eswin,eic7700-clock.h>
 #include "clk_eic7700.h"
 #include "clk.h"
@@ -3713,6 +3715,13 @@ static int eswin_cpu_clk_init(struct platform_device *pdev)
 	int numa_id;
 	char name[128] = { 0 };
 
+	/* Add OPP table from device tree */
+	ret = devm_pm_opp_of_add_table(&pdev->dev);
+	if (ret) {
+		dev_err(dev, "Failed to add OPP table\n");
+		return ret;
+	}
+
 	ret = of_property_read_u32(np, "cpu-default-frequency", &default_freq);
 	if (ret) {
 		dev_info(dev, "cpu-default-frequency not set\n");
diff --git a/drivers/clk/eswin/clk.c b/drivers/clk/eswin/clk.c
index c78c147cf3cf..e33d3405305f 100644
--- a/drivers/clk/eswin/clk.c
+++ b/drivers/clk/eswin/clk.c
@@ -29,6 +29,8 @@
 #include <linux/delay.h>
 #include <linux/util_macros.h>
 #include <linux/gpio/consumer.h>
+#include <linux/devfreq.h>
+#include <linux/pm_opp.h>
 #include <dt-bindings/clock/eswin,eic7700-clock.h>
 #include "clk.h"
 
@@ -317,6 +319,8 @@ static int clk_pll_set_rate(struct clk_hw *hw, unsigned long rate,
 	char clk_cpu_pll_name[50] = { 0 };
 	enum voltage_level cpu_target_voltage = VOLTAGE_0_8V;
 	unsigned long max_rate = rate;
+	struct dev_pm_opp *opp;
+	unsigned long target_volt;
 
 	ret = eswin_calc_pll(&frac_val, &postdiv1_val, &fbdiv_val, &refdiv_val,
 			     (u64)rate, clk);
@@ -388,23 +392,15 @@ static int clk_pll_set_rate(struct clk_hw *hw, unsigned long rate,
 		 * If the board cpu voltage does not support boosting to 0.9V,
 		 * then the frequency cannot exceed 1.6GHz.
 		 */
-		switch (max_rate) {
-		case CLK_FREQ_1800M:
-		case CLK_FREQ_1700M:
-		case CLK_FREQ_1600M:
-		case CLK_FREQ_1500M:
-			cpu_target_voltage = VOLTAGE_0_9V;
-			break;
-		default:
-			cpu_target_voltage = VOLTAGE_0_8V;
-			/*
-			 * For boards that do not support voltage switching,
-			 * the voltage is maintained at 0.8V.
-			 * Therefore, this is also considered successful.
-			 */
-			break;
+		opp = devfreq_recommended_opp(clk->dev, &max_rate, 1);
+		if (IS_ERR(opp)) {
+			return PTR_ERR(opp);
 		}
 
+		target_volt = dev_pm_opp_get_voltage(opp);
+		dev_pm_opp_put(opp);
+		cpu_target_voltage = target_volt > VOLTAGE_0_8V ? VOLTAGE_0_9V:VOLTAGE_0_8V;
+
 		if (clk->cpu_current_voltage !=	cpu_target_voltage) {
 			ret = eswin_clk_set_cpu_voltage(clk->cpu_voltage_gpio,
 							cpu_target_voltage);
@@ -717,6 +713,7 @@ void eswin_clk_register_pll(struct eswin_pll_clock *clks, int nums,
 		init.num_parents = parent_name ? 1 : 0;
 		init.ops = &eswin_clk_pll_ops;
 
+		p_clk->dev = dev;
 		p_clk->id = clks[i].id;
 		p_clk->numa_id = data->numa_id;
 		p_clk->ctrl_reg0 = base + clks[i].ctrl_reg0;
diff --git a/drivers/clk/eswin/clk.h b/drivers/clk/eswin/clk.h
index 48c793ed9c38..9fbe327eb56e 100644
--- a/drivers/clk/eswin/clk.h
+++ b/drivers/clk/eswin/clk.h
@@ -133,12 +133,13 @@ struct eswin_pll_clock {
 };
 
 enum voltage_level {
-	VOLTAGE_0_9V = 900, // Represents 0.9V in millivolts
-	VOLTAGE_0_8V = 800 // Represents 0.8V in millivolts
+	VOLTAGE_0_9V = 900000, // Represents 0.9V in microvolt‌
+	VOLTAGE_0_8V = 800000 // Represents 0.8V in microvolt‌
 };
 
 struct eswin_clk_pll {
 	struct clk_hw hw;
+	struct device *dev;
 	u32 id;
 	int numa_id;
 	void __iomem *ctrl_reg0;
-- 
2.54.0

